torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,167 +10,601 @@
+        }
+        flat_grad = _gather_flat_grad();
+        opt_cond = val(torch::max(flat_grad.abs())) <= tolerance_grad;
+        ls_func_evals = 1;
+      }
     }
+    // update func eval
+    current_evals += ls_func_evals;
+    state.func_evals(state.func_evals() + ls_func_evals);
+
+    // ############################################################
+    // # check conditions
+    // ############################################################
+    if (n_iter == max_iter) break;
+
+    if (current_evals >= *max_eval) break;
+
+    // optimal condition
+    if (opt_cond) break;
+
+    // lack of progress
+    if (val(d.mul(t).abs().max()) <= tolerance_change) break;
+
+    if (abs(loss - prev_loss) < tolerance_change) break;
   }
+
   return orig_loss;
 }
 
+void LBFGS::add_parameters(const std::vector<Tensor>& parameters) {
+  return _add_parameters_new_design(parameters);
+}
+
+const std::vector<Tensor>& LBFGS::parameters() const noexcept {
+  return _parameters_new_design();
+}
+
+std::vector<Tensor>& LBFGS::parameters() noexcept {
+  return _parameters_new_design();
+}
+
+size_t LBFGS::size() const noexcept {
+  return _size_new_design();
+}
+
 void LBFGS::save(serialize::OutputArchive& archive) const {
   serialize(*this, archive);
 }
 
 void LBFGS::load(serialize::InputArchive& archive) {
-  serialize(*this, archive);
+  IValue pytorch_version;
+  if (archive.try_read("pytorch_version", pytorch_version)) {
+    serialize(*this, archive);
+  }
+  else { // deserializing archives saved in old format (prior to version 1.5.0)
+    TORCH_WARN(
+      "Your serialized LBFGS optimizer is still using the old serialization format. "
+      "The func_evals and n_iter value in state will be set to 0, ro will be set to an empty deque "
+      "and al will be set to c10::nullopt because the old LBFGS optimizer didn't save these values."
+      "You should re-save your LBFGS optimizer to use the new serialization format.");
+    Tensor d, t, H_diag, prev_flat_grad, prev_loss;
+    std::deque<Tensor> old_dirs, old_stps;
+    archive("d", d, /*is_buffer=*/true);
+    archive("t", t, /*is_buffer=*/true);
+    archive("H_diag", H_diag, /*is_buffer=*/true);
+    archive("prev_flat_grad", prev_flat_grad, /*is_buffer=*/true);
+    archive("prev_loss", prev_loss, /*is_buffer=*/true);
+    torch::optim::serialize(archive, "old_dirs", old_dirs);
+    torch::optim::serialize(archive, "old_stps", old_stps);
+
+    // NOTE: LBFGS has only global state, but we register it as state for
+    // the first param, because this helps with casting in load_state_dict
+    auto state = std::make_unique<LBFGSParamState>();
+    state->d(d);
+    state->t(t.item<double>());
+    state->H_diag(H_diag);
+    state->prev_flat_grad(prev_flat_grad);
+    state->prev_loss(prev_loss.item<double>());
+    state->old_dirs(old_dirs);
+    state->old_stps(old_stps);
+    state_[c10::guts::to_string(param_groups_.at(0).params().at(0).unsafeGetTensorImpl())] = std::move(state);
```

Comment:
I changed the logic here to only deserialize param state into the first param, because LBFGS param state are global. Let me know if this looks good / needs change :D

Reply:
great catch! thank you :D